File: BOJ/python3/11724.py
'''
연결 요소의 개수
https://www.acmicpc.net/problem/11724
'''
from collections import deque
import sys
import time
start = time.time()  # 시작 시간
sys.stdin = open("input.txt", "r")  # input.txt 파일로 입력


def bfs(node):
    visit[node] = 1
    queue.append(node)
    while len(queue) > 0:
        curr_node = queue.popleft()
        for next_node in adj_list[curr_node]:
            if visit[next_node] == 0:
                visit[next_node] = 1
                queue.append(next_node)


N, M = map(int, input().split())

adj_list = [[] for _ in range(N+1)]
visit = [0]*(N+1)
for _ in range(M):
    u, v = map(int, sys.stdin.readline().split())
    adj_list[u].append(v)
    adj_list[v].append(u)

queue = deque()
ans = 0
for node in range(1, N+1):
    if visit[node] == 0:
        ans += 1
        bfs(node)

        if sum(visit) == N:
            break

print(ans)

# 현재시각 - 시작시간 = 실행 시간(sec)
print("--------------------")
print(f"Elapsed Time: {round(time.time() - start, 3)}s")

File: BOJ/nodejs/11724.js
/*
연결 요소의 개수
https://boj.kr/11724
 */
const start = new Date().getTime();

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .split("\n");

const [N, M] = input[0].split(" ").map(Number); // [정점 수, 간선 수]

const adj = Array.from({ length: N + 1 }, () => new Array(N + 1).fill(0)); // 인접 행렬
for (let m = 0; m < M; m++) {
  const [u, v] = input[1 + m].split(" ").map(Number);
  adj[u][v] = 1;
  adj[v][u] = 1;
}

function bfs(start) {
  const queue = new Array();
  queue.push(start);
  visited[start] = 1;
  while (queue.length > 0) {
    const u = queue.shift();

    for (let v = 1; v < N + 1; v++) {
      if (!adj[u][v] || visited[v]) continue;

      queue.push(v);
      visited[v] = 1;
    }
  }

  return;
}
const visited = new Array(N + 1).fill(0); // 방문 정점
let ans = 0;
for (let v = 1; v < N + 1; v++) {
  if (visited[v]) continue;

  ans++;
  bfs(v);
}

console.log(ans);

const end = new Date().getTime();
console.log("--------------------");
console.log(`Elapsed Time: ${(new Date().getTime() - start) / 1000}s`);
